File: Competitive-Programming-Practice/Random_Practice_Problems/C++/DSU_Implementation.cpp
#include <bits/stdc++.h>

using namespace std;

class DSU {
private:
    vector<int> sz;
    vector<int> parent;
    int num_com;
    int max_size;

public:
    DSU(int n) {
        sz.assign(n + 1, 1);
        iota(parent.begin(), parent.end(), 0);
        num_com = n;
        max_size = 1;
    }
    int find(int i) {
        if (parent[i] == i) {
            return i;
        }
        return parent[i] = find(parent[i]);
    }

    void unite(int a, int b) {
        int root_i = find(a);
        int root_j = find(b);
        if (root_i != root_j) {
            if (sz[root_i] < sz[root_j]) {
                swap(root_i, root_j);
            }
            parent[root_j] = root_i;
            num_com--;
            sz[root_i] += sz[root_j];
            if (sz[root_i] > max_size) {
                max_size = sz[root_i];
            }
        }
    }
};

File: Competitive-Programming-Practice/Random_Practice_Problems/Python/DSU_Implementation.py
class DSU:
    def __init__(self, n):
        sz = [1]*n+1
        parent = list(range(n+1))
        num_com = n
        max_size = 1
    def find(self, i):
        if(self.parent[i] == i):
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]
    def unite(self, i , j):
        root_i = self.find(self.parent[i])
        root_j = self.find(self.parent[j])
        if(root_i != root_j):
            if(root_i > root_j):
                root_i, root_j = root_j, root_i
            self.parent(root_j) = root_i
            self.num_com -= 1
            self.sz[root_i] += self.sz[root_j]
            self.max_size = max(self.sz[root_i], self.max_size)
            